lib/spack/spack/solver/concretize.lp

```diff
@@ -909,6 +1003,105 @@ build_priority(Package, 0)   :- node(Package), not optimize_for_reuse().
 
 #defined installed_hash/2.
 
+%-----------------------------------------------------------------
+% Optimization to avoid errors
+%-----------------------------------------------------------------
+% Some errors are handled as rules instead of constraints because
+% it allows us to explain why something failed. Here we optimize
+% HEAVILY against the facts generated by those rules.
+#minimize{ 0@1000: #true }.
+#minimize{ 1000@1000,Msg: error_conflict_triggered(Msg) }.
```

Comment:
I think if you consolidate the `error_*` rules to a single `error()` predicate as suggested above, we only really need one minimization rule here (at priority 1000), though you may need a few for the different arities.  If you want to preserve priorities ,you can simplify with a `Priority` arg and a fixed number of priorities:

```prolog
#minimize{ 0@1000: #true }.
#minimize{ 0@1001: #true }.
#minimize{ 0@1002: #true }.
#minimize{ 0@1003: #true }.
#minimize{ 0@1004: #true }.
#minimize{ 1@1000+Priority,Message: error(Message, Priority, Arg1) }.
#minimize{ 1@1000+Priority,Message: error(Message, Priority, Arg1, Arg2) }.
#minimize{ 1@1000+Priority,Message: error(Message, Priority, Arg1, Arg2, Arg3) }.
#minimize{ 1@1000+Priority,Message: error(Message, Priority, Arg1, Arg2, Arg3, Arg4) }.
#minimize{ 1@1000+Priority,Message: error(Message, Priority, Arg1, Arg2, Arg3, Arg4, Arg5) }.
```

etc.